src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,510 @@
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper<Document>>>> _cachedDocumentHandlers = new();
+
+        /// <summary>
+        /// Cached handlers of non-document-related messages, indexed by handler message name.
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper<Solution>>>> _cachedWorkspaceHandlers = new();
+
+        private static string GetAssemblyFolderPath(string assemblyFilePath)
+        {
+            return Path.GetDirectoryName(assemblyFilePath)
+                ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+        }
+
+        private async ValueTask<TResult> ExecuteInRemoteOrCurrentProcessAsync<TResult>(
+            Solution? solution,
+            Func<CancellationToken, ValueTask<TResult>> executeInProcessAsync,
+            Func<IRemoteExtensionMessageHandlerService, Checksum?, CancellationToken, ValueTask<TResult>> executeOutOfProcessAsync,
+            CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(_solutionServices, cancellationToken).ConfigureAwait(false);
+            if (client is null)
+                return await executeInProcessAsync(cancellationToken).ConfigureAwait(false);
+
+            if (solution is null)
+            {
+                var result = await client.TryInvokeAsync<IRemoteExtensionMessageHandlerService, TResult>(
+                    (remoteService, cancellationToken) => executeOutOfProcessAsync(remoteService, null, cancellationToken),
+                    cancellationToken).ConfigureAwait(false);
+
+                // If the remote call succeeded, this will have a valid valid in it and can be returned.  If it did not
+                // succeed then we will have already shown the user an error message stating there was an issue making the call,
+                // and it's fine for this to throw again, unwinding the stack up to the caller.
+                return result.Value;
+            }
+            else
+            {
+                var result = await client.TryInvokeAsync<IRemoteExtensionMessageHandlerService, TResult>(
+                    solution,
+                    (remoteService, checksum, cancellationToken) => executeOutOfProcessAsync(remoteService, checksum, cancellationToken),
+                    cancellationToken).ConfigureAwait(false);
+                return result.Value;
+            }
+        }
+
+        private void ClearCachedHandlers_WhileUnderLock()
+        {
+            Contract.ThrowIfTrue(!Monitor.IsEntered(_gate));
+            _cachedWorkspaceHandlers.Clear();
+            _cachedDocumentHandlers.Clear();
+        }
+
+        #region RegisterExtension
+
+        public async ValueTask RegisterExtensionAsync(
```

Comment:
the common pattern is that the service entrypoint immediately calls to the OOPOrInProc helper.  Then the inproc code is right below which actually does the work.